### PR TITLE
Add pd/tikv liveness probe 

### LINKED ISSUE #3799: Liveness and readiness probe in TiDB Operator
## Feature Request

**Is your feature request related to a problem? Please describe:**

Currently the TiDB Operator could be in an out-of-service state, such as waiting for the next leader election lease. This can not be easily monitored now.

**Describe the feature you'd like:**

Add liveness and readiness probe as defined in [Kubernetes Pod Lifecycle](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#when-should-you-use-a-liveness-probe).

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### LINKED ISSUE #3857: Add readiness probe and liveness probe for the components
## Feature Request

**Is your feature request related to a problem? Please describe:**

Currently, we do not support setting the readiness probe and liveness probe for the components except for the readiness probe for TiDB, however, we should support setting them for the following reasons:

- The readiness probe will also be honored by the sts controller during the rolling update because sts controller will wait for the Pod ready until updating the next Pod.
- Some Pods can hang there and can be recovered by the Liveness probe

The following cases should be considered:

- TLS is enabled for all the components and between TiDB server and client, because the access to the status port of one component may also need certs like TiDB
- The probe configuration should be able to customize since in some scenarios we may have to customize them

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by mikechengwei**
### What problem does this PR solve?

Fix #3799 .
Fix #3857 .
### What is changed and how does it work?
I find there is no curl command in  tidb 4.0.9 or later image , so temporarily use the tcp port method.
### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```

### CODE REVIEW

**Comment by KanShiori** [comment]
PTAL @hanlins @charleszheng44

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 1]
```diff
@@ -911,11 +922,11 @@ const (
...
 )
 
-// TiDBProbe contains details of probing tidb.
+// Probe contains details of probing tidb.
 // +k8s:openapi-gen=true
-// default probe by TCPPort on 4000.
-type TiDBProbe struct {
-	// "tcp" will use TCP socket to connetct port 4000
+// default probe by TCPPort on tidb 4000 / tikv 20160 / pd 2349.
+type Probe struct {
```
Comment (hanlins) [inline pkg/apis/pingcap/v1alpha1/types.go:928]: Would be great if we can include an HTTPGet type handler for the probe.

File: pkg/manager/member/tikv_member_manager.go [hunk 2]
```diff
@@ -1025,6 +1041,15 @@ func tikvStatefulSetIsUpgrading(podLister corelisters.PodLister, pdControl pdapi
 	return false, nil
 }
 
+// TODO: Support check pd status http request in future.
```
Comment (hanlins) [inline pkg/manager/member/tikv_member_manager.go:1044]: nit
```suggestion
// TODO: Support check tikv status http request in future.
```

File: pkg/manager/member/pd_member_manager.go [hunk 3]
```diff
@@ -951,6 +968,15 @@ func (m *pdMemberManager) collectUnjoinedMembers(tc *v1alpha1.TidbCluster, set *
 	return nil
 }
 
+// TODO: Support check status http request in future.
+func buildPDReadinessProbHandler() corev1.Handler {
```
Comment (hanlins) [inline pkg/manager/member/pd_member_manager.go:972]: Can we include `*v1alpha1.TidbCluster` as an input parameter here although it's not in use for now? Just like for `TiDB`, we might need to construct some non-default handler based on the `TidbCluster` spec later.

File: pkg/manager/member/tikv_member_manager.go [hunk 4]
```diff
@@ -1025,6 +1041,15 @@ func tikvStatefulSetIsUpgrading(podLister corelisters.PodLister, pdControl pdapi
...
+func buildTiKVReadinessProbHandler() corev1.Handler {
```
Comment (hanlins) [inline pkg/manager/member/tikv_member_manager.go:1045]: Similar to `buildTiDBReadinessProbHandler`, maybe we need to introduce an `*v1alpha1.TidbCluster` input parameter here for future use.

**Review by hanlins — COMMENTED** [review]
Hi @mikechengwei, thanks for your contribution! Have some minor comments but overall looks promising. One thing is that for PD we probably want to introduce an HTTP-based readiness probe in the future (ref https://github.com/tikv/pd/pull/5685#issuecomment-1310950417), so it would be great to enable HTTP typed probe here. I don't mind resolving that in a follow-up PR 😄

**Comment by mikechengwei** [comment]
LGTM

**Comment by mikechengwei** [comment]
@charleszheng44 Do you have time help me review this pr ? Thanks a lot.

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 5]
```diff
@@ -616,6 +622,11 @@ type TiKVSpec struct {
 	// ScalePolicy is the scale configuration for TiKV
 	// +optional
 	ScalePolicy ScalePolicy `json:"scalePolicy,omitempty"`
+
+	// ReadinessProbe describes actions that probe the tikv's readiness.
+	// the default behavior is like setting type as "tcp"
+	// +optional
+	ReadinessProbe *Probe `json:"readinessProbe,omitempty"`
```
Comment (KanShiori) [inline pkg/apis/pingcap/v1alpha1/types.go:629]: how about putting it in the `ComponentSpec`

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 6]
```diff
@@ -897,10 +898,6 @@ type TiDBSpec struct {
 	// Defaults to Kubernetes default storage class.
 	// +optional
 	StorageClassName *string `json:"storageClassName,omitempty"`
-	// ReadinessProbe describes actions that probe the tidb's readiness.
-	// the default behavior is like setting type as "tcp"
-	// +optional
-	ReadinessProbe *TiDBProbe `json:"readinessProbe,omitempty"`
```
Comment (KanShiori) [inline pkg/apis/pingcap/v1alpha1/types.go:903]: just to confirm, won't cause a rolling upgrade after putting it into `ComponentSpec`, right?

**Review by KanShiori — APPROVED** [review]

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 6 above)
Comment (csuzhangxc) [reply pkg/apis/pingcap/v1alpha1/types.go:903]: @mikechengwei have you tested this?

**Comment by csuzhangxc** [comment]
`ci / pull (e2e-examples)` seems always failed

**Comment by mikechengwei** [comment]
> `ci / pull (e2e-examples)` seems always failed

re-run successfully.

**Pull request merged**